## Numbers

Generate a dictionary where the keys are the numbers between 1 and 20 and the values are the square of the keys.

In [1]:
# Option 1: "dictionary comprehension"
numbers = {n:n**2 for n in range(1,21)}
print(numbers)

In [2]:
# Option 2
numbers = {}
for n in range(1,21):
    numbers[n] = n**2
numbers

{1: 1,
 2: 4,
 3: 9,
 4: 16,
 5: 25,
 6: 36,
 7: 49,
 8: 64,
 9: 81,
 10: 100,
 11: 121,
 12: 144,
 13: 169,
 14: 196,
 15: 225,
 16: 256,
 17: 289,
 18: 324,
 19: 361,
 20: 400}

The result is

    {1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400}


Update the dictionary from above with the given dictionary *additional_dict*

In [3]:
additional_dict = {40: 1600, 41: 1681, 42: 1764}

numbers.update(additional_dict)

numbers

{1: 1,
 2: 4,
 3: 9,
 4: 16,
 5: 25,
 6: 36,
 7: 49,
 8: 64,
 9: 81,
 10: 100,
 11: 121,
 12: 144,
 13: 169,
 14: 196,
 15: 225,
 16: 256,
 17: 289,
 18: 324,
 19: 361,
 20: 400,
 40: 1600,
 41: 1681,
 42: 1764}

The result is

    {1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100, 11: 121, 12: 144, 13: 169, 14: 196, 15: 225, 16: 256, 17: 289, 18: 324, 19: 361, 20: 400, 40: 1600, 41: 1681, 42: 1764}


## NYC marathon winners

Here's a (nested) list of the female NYC marathon winners from wikipedia. https://en.wikipedia.org/wiki/List_of_winners_of_the_New_York_City_Marathon in the format

            [['Year 1', 'Winner 1, 'Country 1', 'Time 1', 'Notes 1'],
             ['Year 2', 'Winner 2, 'Country 2', 'Time 2', 'Notes 2'], 
             ...]]
The follwing cell reads in the list from the file *female_winners.csv* and modifies the list to match the format above. Reading in the file will be covered later.

In [4]:
with open("female_winners.csv") as f:
    winners = f.readlines()

winners = [entry.replace("\n","").split(",") for entry in winners]
winners = winners[1:]
winners[:3]

[['1971', 'Beth Bonner', 'United States', '2:55:22', 'World record'],
 ['1972', 'Nina Kuscsik', 'United States', '3:08:41', ''],
 ['1973', 'Nina Kuscsik', 'United States', '2:57:07', 'Second victory']]

### Winner count
Use a dictionary to count how often each winner won the marathon.

*Hint: Loop through the list. Start with an empty dictionary and add winners as a key-value pair as they appear, so you'll end up with a dictionary like {"winner1":count1, "winner2":count2, ...}.*

In [5]:
winner_count = {}
for entry in winners:
    winner = entry[1]
    if winner in winner_count:
        winner_count[winner] += 1
    else:
        winner_count[winner] = 1
winner_count

{'Beth Bonner': 1,
 'Nina Kuscsik': 2,
 'Kathrine Switzer': 1,
 'Kim Merritt': 1,
 'Miki Gorman': 2,
 'Grete Waitz': 9,
 'Allison Roe': 1,
 'Priscilla Welch': 1,
 'Ingrid Kristiansen': 1,
 'Wanda Panfil': 1,
 'Liz McColgan': 1,
 'Lisa Ondieki': 1,
 'Uta Pippig': 1,
 'Tegla Loroupe': 2,
 'Anuța Cătună': 1,
 'Franziska Rochat-Moser': 1,
 'Franca Fiacconi': 1,
 'Adriana Fernández': 1,
 'Lyudmila Petrova': 1,
 'Margaret Okayo': 2,
 'Joyce Chepchumba': 1,
 'Paula Radcliffe': 3,
 'Jeļena Prokopčuka': 2,
 'Derartu Tulu': 1,
 'Edna Kiplagat': 1,
 'Firehiwot Dado': 1,
 'Priscah Jeptoo': 1,
 'Mary Keitany': 4,
 'Shalane Flanagan': 1,
 'Joyciline Jepkosgei': 1}

The result is

    {'Beth Bonner': 1,
     'Nina Kuscsik': 2,
     'Kathrine Switzer': 1,
     'Kim Merritt': 1,
     'Miki Gorman': 2,
     'Grete Waitz': 9,
     'Allison Roe': 1,
     'Priscilla Welch': 1,
     'Ingrid Kristiansen': 1,
     'Wanda Panfil': 1,
     'Liz McColgan': 1,
     'Lisa Ondieki': 1,
     'Uta Pippig': 1,
     'Tegla Loroupe': 2,
     'Anuța Cătună': 1,
     'Franziska Rochat-Moser': 1,
     'Franca Fiacconi': 1,
     'Adriana Fernández': 1,
     'Lyudmila Petrova': 1,
     'Margaret Okayo': 2,
     'Joyce Chepchumba': 1,
     'Paula Radcliffe': 3,
     'Jeļena Prokopčuka': 2,
     'Derartu Tulu': 1,
     'Edna Kiplagat': 1,
     'Firehiwot Dado': 1,
     'Priscah Jeptoo': 1,
     'Mary Keitany': 4,
     'Shalane Flanagan': 1,
     'Joyciline Jepkosgei': 1}

### Structure one sublist

Create a dictionary like

    {"winner":value1, "country":value2, "time":value3, "note":value4}

with for the 0th item of the winner list.

*Hint: Start with an empty dictionary. Use the zip() function and the list ["winner","country","","note"]*

In [6]:
# Option 1
winner_0 = winners[0][1:]

winner_dict_0 = {}
for key, value in zip( ["winner","country","time","note"], winner_0):
    winner_dict_0[key] = value
    
winner_dict_0

{'winner': 'Beth Bonner',
 'country': 'United States',
 'time': '2:55:22',
 'note': 'World record'}

In [7]:
# Option 2
winner_0 = winners[0][1:]
winner_dict_0 = {key:value for key, value in zip( ["winner","country","time","note"], winner_0)}
winner_dict_0

{'winner': 'Beth Bonner',
 'country': 'United States',
 'time': '2:55:22',
 'note': 'World record'}

The result is 

    {'winner': 'Beth Bonner',
    'country': 'United States',
    'time': '2:55:22',
    'note': 'World record'}

### Structure the whole list, nested dictionary

Now, create a nested dictionary like

    {year1: {"winner":value1, "country":value2, "time":value3, "note":value4},
     year2: {"winner":value1, "country":value2, "time":value3, "note":value4},
     ...}
     
*Hint: Start with an empty dictionary. Use a nested for loop.

*Hint 2: Some structure for the nested loop is already given below. Make sure to create a new empty dictionary for each inner dictionary and not just overwrite the keys. More explanations on why this is important will be given in the solutions.

In [8]:
winner_dict = {}
for item in winners:
    year = item[0]
    winner_i = item[1:]
    winner_dict_i = {}
    for key, value in zip( ["winner","country","time","note"], winner_i):
        winner_dict_i[key] = value
    winner_dict[year] = winner_dict_i
winner_dict

{'1971': {'winner': 'Beth Bonner',
  'country': 'United States',
  'time': '2:55:22',
  'note': 'World record'},
 '1972': {'winner': 'Nina Kuscsik',
  'country': 'United States',
  'time': '3:08:41',
  'note': ''},
 '1973': {'winner': 'Nina Kuscsik',
  'country': 'United States',
  'time': '2:57:07',
  'note': 'Second victory'},
 '1974': {'winner': 'Kathrine Switzer',
  'country': 'United States',
  'time': '3:07:29',
  'note': ''},
 '1975': {'winner': 'Kim Merritt',
  'country': 'United States',
  'time': '2:46:14',
  'note': 'Course record'},
 '1976': {'winner': 'Miki Gorman',
  'country': 'United States',
  'time': '2:39:11',
  'note': 'Course record'},
 '1977': {'winner': 'Miki Gorman',
  'country': 'United States',
  'time': '2:43:10',
  'note': 'Second victory'},
 '1978': {'winner': 'Grete Waitz',
  'country': 'Norway',
  'time': '2:32:30',
  'note': 'World record'},
 '1979': {'winner': 'Grete Waitz',
  'country': 'Norway',
  'time': '2:27:33',
  'note': '"World record'},
 '1980'

The first part of the resulting dictionary is

    {'1971': {'winner': 'Beth Bonner',
              'country': 'United States',
              'time': '2:55:22',
              'note': 'World record'},
     '1972': {'winner': 'Nina Kuscsik',
              'country': 'United States',
              'time': '3:08:41',
              'note': ''},
     '1973': {'winner': 'Nina Kuscsik',
              'country': 'United States',
              'time': '2:57:07',
              'note': 'Second victory'},
     '1974': {'winner': 'Kathrine Switzer',
              'country': 'United States',
              'time': '3:07:29',
              'note': ''},
       ... }

Explanation why it is important to create an empty dictionary for each inner dictionary and not just overwrite the keys:

If you don't do that, the following happens:

In [9]:
winner_dict = {}
winner_dict_i = {}   # create an empty dictionary for the inner ones just once and overwrite the keys in the loop:
for item in winners:
    year = item[0]
    winner_i = item[1:]
    for key, value in zip( ["winner","country","time","note"], winner_i):
        winner_dict_i[key] = value

    winner_dict[year] = winner_dict_i.copy()
    print("Memory ID of winner_dict[year]:", id(winner_dict[year]))

winner_dict

Memory ID of winner_dict[year]: 139700714369984
Memory ID of winner_dict[year]: 139700723354432
Memory ID of winner_dict[year]: 139700714371424
Memory ID of winner_dict[year]: 139700714371024
Memory ID of winner_dict[year]: 139700723179600
Memory ID of winner_dict[year]: 139700723252352
Memory ID of winner_dict[year]: 139700714371664
Memory ID of winner_dict[year]: 139700714371744
Memory ID of winner_dict[year]: 139700714371824
Memory ID of winner_dict[year]: 139700714371904
Memory ID of winner_dict[year]: 139700714371984
Memory ID of winner_dict[year]: 139700714372064
Memory ID of winner_dict[year]: 139700714372144
Memory ID of winner_dict[year]: 139700714372224
Memory ID of winner_dict[year]: 139700714372304
Memory ID of winner_dict[year]: 139700714372384
Memory ID of winner_dict[year]: 139700714372464
Memory ID of winner_dict[year]: 139700714372544
Memory ID of winner_dict[year]: 139700714372624
Memory ID of winner_dict[year]: 139700714372704
Memory ID of winner_dict[year]: 13970071

{'1971': {'winner': 'Beth Bonner',
  'country': 'United States',
  'time': '2:55:22',
  'note': 'World record'},
 '1972': {'winner': 'Nina Kuscsik',
  'country': 'United States',
  'time': '3:08:41',
  'note': ''},
 '1973': {'winner': 'Nina Kuscsik',
  'country': 'United States',
  'time': '2:57:07',
  'note': 'Second victory'},
 '1974': {'winner': 'Kathrine Switzer',
  'country': 'United States',
  'time': '3:07:29',
  'note': ''},
 '1975': {'winner': 'Kim Merritt',
  'country': 'United States',
  'time': '2:46:14',
  'note': 'Course record'},
 '1976': {'winner': 'Miki Gorman',
  'country': 'United States',
  'time': '2:39:11',
  'note': 'Course record'},
 '1977': {'winner': 'Miki Gorman',
  'country': 'United States',
  'time': '2:43:10',
  'note': 'Second victory'},
 '1978': {'winner': 'Grete Waitz',
  'country': 'Norway',
  'time': '2:32:30',
  'note': 'World record'},
 '1979': {'winner': 'Grete Waitz',
  'country': 'Norway',
  'time': '2:27:33',
  'note': '"World record'},
 '1980'

In [10]:
winner_dict = {}
winner_dict_i = {}   # create an empty dictionary for the inner ones just once and overwrite the keys in the loop:
for item in winners:
    year = item[0]
    winner_i = item[1:]
    for key, value in zip( ["winner","country","time","note"], winner_i):
        winner_dict_i[key] = value
    
    print("Memory ID of winner_dict_i:", id(winner_dict_i))
    winner_dict[year] = winner_dict_i.copy()
    print("Memory ID of winner_dict[year]:", id(winner_dict[year]))
winner_dict

Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723399568
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723236048
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723236208
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723233408
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723235888
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723235328
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723235088
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723234768
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723234688
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 139700723234608
Memory ID of winner_dict_i: 139700723233168
Memory ID of winner_dict[year]: 1397

{'1971': {'winner': 'Beth Bonner',
  'country': 'United States',
  'time': '2:55:22',
  'note': 'World record'},
 '1972': {'winner': 'Nina Kuscsik',
  'country': 'United States',
  'time': '3:08:41',
  'note': ''},
 '1973': {'winner': 'Nina Kuscsik',
  'country': 'United States',
  'time': '2:57:07',
  'note': 'Second victory'},
 '1974': {'winner': 'Kathrine Switzer',
  'country': 'United States',
  'time': '3:07:29',
  'note': ''},
 '1975': {'winner': 'Kim Merritt',
  'country': 'United States',
  'time': '2:46:14',
  'note': 'Course record'},
 '1976': {'winner': 'Miki Gorman',
  'country': 'United States',
  'time': '2:39:11',
  'note': 'Course record'},
 '1977': {'winner': 'Miki Gorman',
  'country': 'United States',
  'time': '2:43:10',
  'note': 'Second victory'},
 '1978': {'winner': 'Grete Waitz',
  'country': 'Norway',
  'time': '2:32:30',
  'note': 'World record'},
 '1979': {'winner': 'Grete Waitz',
  'country': 'Norway',
  'time': '2:27:33',
  'note': '"World record'},
 '1980'